aiven/kafka_topic_availability.go

```diff
@@ -0,0 +1,73 @@
+package aiven
+
+import (
+	"fmt"
+	"github.com/aiven/terraform-provider-aiven/pkg/cache"
+	"log"
+	"time"
+
+	"github.com/aiven/aiven-go-client"
+	"github.com/hashicorp/terraform/helper/resource"
+)
+
+// KafkaTopicAvailabilityWaiter is used to refresh the Aiven Kafka Topic endpoints when
+// provisioning.
+type KafkaTopicAvailabilityWaiter struct {
+	Client      *aiven.Client
+	Project     string
+	ServiceName string
+	TopicName   string
+}
+
+// RefreshFunc will call the Aiven client and refresh it's state.
+func (w *KafkaTopicAvailabilityWaiter) RefreshFunc() resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		topicCache := cache.GetTopicCache()
+		topic, ok := topicCache.LoadByTopicName(w.Project, w.ServiceName, w.TopicName)
+
+		if !ok {
+			list, err := w.Client.KafkaTopics.List(w.Project, w.ServiceName)
+
+			if err != nil {
+				aivenError, ok := err.(aiven.Error)
+				// Topic creation is asynchronous so it is possible for the creation call to
+				// have completed successfully yet fetcing topic info fails with 404.
+				if ok && aivenError.Status == 404 {
+					return nil, "CONFIGURING", nil
+				}
+				// Getting topic info can sometimes temporarily fail with 501. Don't treat
+				// that as fatal error but keep on retrying instead
+				if ok && aivenError.Status == 501 {
```

Comment:
Not changed in this pr but this should probably handle 502 the same way as 501, I think the operation can sometimes fail with that error as well.